dbms/src/DataTypes/DataTypeFactory.cpp

```diff
@@ -40,22 +40,44 @@ DataTypePtr DataTypeFactory::get(const String & full_name) const
     ASTPtr ast = parseQuery(parser, full_name.data(), full_name.data() + full_name.size(), "data type", 0);
     return get(ast);
 }
+// DataTypeFactory is a Singleton, so need to be protected by lock.
+DataTypePtr DataTypeFactory::getOrSet(const String & full_name)
+{
+    {
+        std::shared_lock lock(rw_lock);
+        auto it = fullname_types.find(full_name);
```

Comment:
What is the full_name of a DataTypeEnum? Will it contains the enum values?

Reply:
> What is the full_name of a DataTypeEnum? Will it contains the enum values?

It will contains the enum values, so that it will generate too many `full_name`. I think we can skip inserting  DataTypeEnum into `full_name` when the capacity limit is almost reached, which ensures that most datatypes can be cached.